pkg/resource/pipe/hooks.go

```diff
@@ -0,0 +1,990 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package pipe
+
+import (
+	"fmt"
+	"reflect"
+	"time"
+
+	ackcompare "github.com/aws-controllers-k8s/runtime/pkg/compare"
+	ackrequeue "github.com/aws-controllers-k8s/runtime/pkg/requeue"
+	svcsdk "github.com/aws/aws-sdk-go/service/pipes"
+
+	svcapitypes "github.com/aws-controllers-k8s/pipes-controller/apis/v1alpha1"
+)
+
+const (
+	defaultRequeueDelay = time.Second * 5
+)
+
+var (
+	requeueWaitWhileCreating = ackrequeue.NeededAfter(
+		fmt.Errorf("pipe in status %q, requeueing", svcsdk.PipeStateCreating),
+		defaultRequeueDelay,
+	)
+
+	requeueWaitWhileUpdating = ackrequeue.NeededAfter(
+		fmt.Errorf("pipe in status %q, cannot be modified or deleted", svcsdk.PipeStateUpdating),
+		defaultRequeueDelay,
+	)
+
+	requeueWaitWhileDeleting = ackrequeue.NeededAfter(
+		fmt.Errorf("pipe in status %q, cannot be modified or deleted", svcsdk.PipeStateDeleting),
+		defaultRequeueDelay,
+	)
+)
+
+// hasNilDifference returns true if the supplied subjects' nilness is
+// different
+func hasNilDifference(a, b interface{}) bool {
+	if isEmpty(a) || isEmpty(b) {
+		if (isEmpty(a) && isNotEmpty(b)) || (isEmpty(b) && isNotEmpty(a)) {
+			return true
+		}
+	}
+	return false
+}
+
+// isEmpty checks the passed interface argument for Nil or empty struct value
+// (with zero values). For interfaces, only 'i==nil' check is not sufficient.
+// https://tour.golang.org/methods/12 More details:
+// https://mangatmodi.medium.com/go-check-nil-interface-the-right-way-d142776edef1
+func isEmpty(i interface{}) bool {
+	if i == nil {
+		return true
+	}
+
+	switch reflect.TypeOf(i).Kind() {
+	case reflect.Ptr, reflect.Map, reflect.Array, reflect.Chan, reflect.Slice:
+		if reflect.ValueOf(i).IsNil() {
+			return true
+		}
+
+		if reflect.ValueOf(i).Elem().Kind() == reflect.Struct {
+			return reflect.ValueOf(i).Elem().IsZero()
+		}
+	}
+	return false
+}
+
+func isNotEmpty(i interface{}) bool {
+	return !isEmpty(i)
+}
+
+func customPreCompare(
```

Comment:
I guess this is fine for now, but as you know, this is fragile and requires a lot of effort to maintain (versus generated code). I think in the future, what we should do is provide a hook to *replace the definition of `hasNilDifference` with a custom function. That way, we could rely on code generation for all of this.

Reply:
Exactly! Would love that too